### PR TITLE
Create config

### PR DESCRIPTION
**Pull request by GeoWill**
This is laying the ground work for the `create-config` command.

At the moment the functions to get basic information from the shapefile are kept separate from the class `BoundaryConfig()` since that class is to represent a JSON object in the config file,  so it didn't seem right to give it behaviors that involve reading shapefiles. 

Next steps are:
* Make `bin/create-config` script which will behave as described in the readme. 
* Perhaps some functional tests.

### CODE REVIEW

File: bin/get-report.r
Comment (mhl) [inline bin/get-report.r]: As I mentioned in Slack, this commit introduces a typo here.

File: commons_boundaries/lib.py
Comment (mhl) [inline commons_boundaries/lib.py]: I think you can simplify this to `return to_string(fiona.open(str(shp_path)).crs)` - if the `.prj` file is missing it looks like `fiona.open(str(shp_path)).crs` returns `{}` and `to_string` turns that into `''`

File: commons_boundaries/lib.py
Comment (mhl) [inline commons_boundaries/lib.py]: It's a very minor style thing, but I think I'd drop the explicit else in cases like this, so the shape is:

```
if foo:
    return 'bar'
return ''
```

File: tests/test_boundary_config.py
Comment (mhl) [inline tests/test_boundary_config.py]: If this is a class method, then the first parameter will be the class, not an instance of the class, so it's conventional to call it `cls` rather than `self`.

There's also a small thing that I think I alluded to this before, but not explicitly - I'd tend to use `setUp` unless there's a really good reason that you have to use `setUpClass`, so it's harder to break test isolation (i.e. have the actions of one test method unexpectedly affect another).

File: commons_boundaries/boundary_config.py
Comment (mhl) [inline commons_boundaries/boundary_config.py]: I think if the docstring isn't going to be more descriptive than the class name, it's probably best to just leave it out ;)

File: tests/test_lib.py
Comment (mhl) [inline tests/test_lib.py]: I think my comments elsewhere about trying to avoid using `setUpClass` (or changing the name of the `self` parameter if you're not) apply here too, though it's not a big deal.

File: tests/test_lib.py
Comment (mhl) [inline tests/test_lib.py]: I was going to comment on the tiny point that the `./` isn't necessary here, but perhaps it could be made more robust overall by make the path relative to the current file, like:

```
tests_data = Path(__file__).parent / 'data'
self.shp_path = tests_data / 'italy-source' / 'CAMERA_PLURI_2017.shp'
self.no_meta_shp_path  = tests_data / 'no_meta' / 'CAMERA_PLURI_2017.shp'
```

If someone moves the test file or the example shapefiles you'd still need to update this, but it should be more obvious what the problem is.

File: tests/test_lib.py
Comment (mhl) [inline tests/test_lib.py]: Since `tearDownClass` is empty, you could omit it.

File: tests/functional_test.py
Comment (mhl) [inline tests/functional_test.py]: It makes me a bit nervous changing directory in the `setUp` and `tearDown` methods here, since that's global state for the process. You can make this more localized by using the `cwd` argument to the subprocess command instead.

File: tests/functional_test.py
Comment (mhl) [inline tests/functional_test.py]: When you see nearly-repeated lines like this, I think it's good to think about ways to reduce the repetition. Even though it's not very important in this particular case, it's good to get used to keeping the DRY (don't repeat yourself) principle in mind. e.g. here this could be:

```
for ext in ('.shp', '.shx', '.cpg', '.dbf', '.prj'):
    shutil.copy(str(camera.with_suffix(ext)), str(self.src_path))
```

File: bin/create-config
Comment (mhl) [inline bin/create-config]: If you're using the `/` operator to build paths in most places, I'd be consistent and do it everywhere. e.g. here:

```
boundary_path = Path('boundaries') / 'build' / args.boundary
```

(n.b. I also removed the unneeded `Path` around `args.boundary`)

File: bin/create-config
Comment (mhl) [inline bin/create-config]: I think this is potentially a bit confusing - I think you've added this so that if there's a zero-byte JSON file there, it's treated as a JSON file of `[]`. But this will also catch any syntax error in the JSON, and treat it as if there's nothing in the file without producing an error, and actually overwrite the file with a syntax error losing anything that was previously in it.

I think we should assume that if the `.json` file exists at all, it should be well-formed JSON, and otherwise produce an error. (This will mean changing your test for this behaviour so that it doesn't `.touch()` the file.)

You can still handle the file being missing as "no config data" with something like the following:

```python
try:
    with open('boundaries/boundary-config.json') as f:
        config_entries = json.load(f)
except FileNotFoundError:
    config_entries = []
```

File: bin/create-config
Comment (mhl) [inline bin/create-config]: This will create a strange file, where your JSON is an array where each element is serialized JSON. This will make it generally hard to understand. Instead, I'd change your `dumps` method on `BoundaryConfig` into an `as_dict` method which returns the `OrderedDict`, not serialized JSON.

File: bin/create-config
Comment (mhl) [inline bin/create-config]: A tiny stylistic thing, but I'd try to keep all variable names as snake case, rather than lower camel case.

**Review by mhl — CHANGES_REQUESTED** [review]
I think this is on the right track, but I've left some (I hope helpful) comments inline. The most significant ones are about how the JSON is serialized by `create_config`. I hope that makes sense.

One other thing I should point out is that generally we don't put the failing tests in one commit and the implementation that makes it pass in another - see the last point here: http://mysociety.github.io/coding-standards.html#tests - the tests and the implementation should generally be in the same commit.

Another really minor thing, but worth getting into the habit of fixing, is that the subject line of the commit ec72e190af2bda2 is too long (the same coding standards document I linked to above has some advice about formatting commit messages).

Anyway, this is nearly there, I think!

File: bin/create-config
Comment (mhl) [reply bin/create-config]: Another thing to think about here is if you want to detect that there's an existing entry for the boundary you're building? (Genuine question: I haven't thought about if you want to handle that specially or not.)

**Comment by GeoWill** [comment]
I'm working through the suggestions, and submitting fixups.

So as not to loose the comments by changing the commit history I'll put a couple of 'todos' below which need to be done before merging, but after approval.
- [ ] Squash failing tests with the code that makes them pass.
- [ ] Shorten first lines of commits where they're over 72 chars.

File: commons_boundaries/lib.py
Comment (mhl) [inline commons_boundaries/lib.py:23]: I don't think you need to change this for this PR, but I think it's worth mentioning that the pattern here (functions with a common parameter in all of them) and the naming of the functions, makes me think that this could be a `class` which is created with `shp_path`, and has methods: `proj4(self)`, `encoding(self)` and `write_csv(self, output_csv_path)`. So I'm mentioning this as something to consider the next time you're looking at this code.

**Review by mhl — CHANGES_REQUESTED** [review]
There are a couple of my comments from the first review that I think you've missed - if you step through the PR commit-by-commit GitHub should show you them, I think.

The tests don't pass at the moment, but that's just because you've left in an `ipdb.set_trace()`, so it'd be good to remove that.

I've reservations about how the `method` parameter might be used, but I think we can deal with that better down the line, when more of this pipeline is in place.

I assume you're wanting to start using this soon, so do go ahead and merge this after fixing those 3 things :+1:

**Pull request merged**
